### PR TITLE
Aim assists

### PR DESCRIPTION
**Pull request by Fueredoriku**
- Tweaked player hitboxes a bit
- Made gamepads have larger bullet hit radius than keyboards
- Fix aiming edge case once and for all
- Made bullets actually hit stuff even if the gun output is inside a wall
   -  Note: Can still shoot through stuff if the gun fully penetrates a mesh and the output ends up on the other side, I think we should keep it like this for now
   - Image example: image

### CODE REVIEW

**Review by toberge — CHANGES_REQUESTED** [review]
Seems to work fine for hitting players that are some distance away, however:

- At close ranges the aim assist might be too aggressive.
- Since it aims in on _anything_ and works a tad too well at close range, it might just as well lead to a player's demise when they aim at another player from behind an explosive barrel.
- This implementation only targets hitscan bullets and effectively makes the bullets way larger than they are. I'd imagined that the aim assist would happen when the aim target is set in PlayerManager, which would make it look like the shot lands where it actually does, though the angle might look a little odd to the player. I'm not sure how other FPS games do this though.
- This implementation does at least seem inflexible for future non-hitscan projectiles. At present it shouldn't matter since the current positional projectiles are so chaotic/slow/huge that they'd hit or not hit independently of what happens.
- Making lines hit like cylinders also greatly buffs the coil barrel, since the zigzag pattern doesn't stray _that_ far from the center. I haven't tested enough to determine just how much of a buff it is, but it seems _very_ significant.
- It felt very odd to see bullet holes appear on a wall I wasn't even aiming at, perhaps we could prioritize players (and _some_ other hitboxes) exclusively for the aim assist, so that hits don't land on a wall that is real close when they should've hit a player. These screenshots show just how ridiculously hard this makes it to hit players hiding slightly behind an obstacle:  
  image  
  
image

 

I also left some comments on the code, though most of them might be obsolete if you actually rework how the aim assist works.

And do note that "the aiming edgecase" is quite vague – you mean the self-damage edgecase here, while the gun-inside-object edgecase is still a thing.

File: Assets/Scripts/Augment/ProjectileController.cs [hunk 1]
```diff
@@ -52,6 +52,9 @@ public class ProjectileState
     // TODO: Make this to anything
     public float size = 0f;
```
[the remark this answers is not in the document]
Comment (toberge) [reply Assets/Scripts/Augment/ProjectileController.cs]: (literally just remove it as size is actually used for something)

**Review by toberge — CHANGES_REQUESTED** [review]
Works much better than the previous approach! I still have quite a few concerns, though none as major as last time:

- Hats get stuck in what looks like thin air
- If you kill a player with the hat barrel, the hats that stuck to the aim assist hitbox will hang around in midair, as that hitbox isn't parented to the player
- These hitboxes are even easier to hit at close range, which might not be ideal
- You do have the odd look of shots hitting thin air and somehow the opponent takes damage (though for non-magnetism aim assist it's either this or having the shot angle towards the player that is hit)
- The cannon barrel got massively buffed by this too
- Actually, the assisting hitboxes may be a little too large in the first place :sweat_smile:
  - Looking at it more closely, it gives way more lenience horizontally than vertically  – especially when the players are in profile:
  
image
  image

**Comment by Fueredoriku** [comment]
image
Hats getting stuck in air is really a separate issue, which while exerberated in this PR, is really a symptom of a seperate issue #164 .
Adressing this issue here is not something I want to do.

File: Assets/Scripts/Augment/ProjectileController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/ProjectileController.cs:52]: Removed

File: Assets/Scripts/Augment/BulletController.cs [hunk 2]
```diff
@@ -107,18 +111,18 @@ private void FireProjectile()
...
-                    OnColliderHit?.Invoke(collider, ref projectile);
                     if (hitbox != null)
                     {
+                        if (hitbox.health.Player == player && projectile.distanceTraveled < player.GunController.OutputTransitionDistance)
+                            continue;
                         OnHitboxCollision?.Invoke(hitbox, ref projectile);
                     }
-
+                    OnColliderHit?.Invoke(collider, ref projectile);
+                    projectile.position = collisions[0].point;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/BulletController.cs]: Forgot to move, is now moved

File: Assets/Scripts/Augment/ProjectileController.cs [hunk 3]
```diff
@@ -95,6 +95,9 @@ public abstract class ProjectileController : MonoBehaviour
     [HideInInspector]
     public PlayerManager player;
 
+    // Checks the extra large Hitboxes intended for aim assist
+    public bool HitAssist = false;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/ProjectileController.cs:98]: Removed

File: Assets/Scripts/Augment/GunController.cs [hunk 4]
```diff
@@ -32,6 +33,9 @@ public class GunController : MonoBehaviour
     // Inputs
     public bool triggerHeld, triggerPressed;
     public Vector3 target;
+    public bool TargetIsTooClose;
+    // Extra margin for hit detection intended for aim assist
+    public bool HitAssist = false;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/GunController.cs]: removed

File: Assets/Scripts/Augment/BulletController.cs [hunk 5]
```diff
@@ -23,6 +23,8 @@ public class BulletController : ProjectileController
     private VisualEffect trail;
     public GameObject Trail => trail.gameObject;
 
+    private bool hitAssist = false;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/BulletController.cs]: removed

**Review by toberge — APPROVED** [review]
Code looks good and the aim assist seems to work fine now :+1:  (though I noticed it being less effective when the other player was partially obscured by a metal pillar in grand canyon (may have just been the shape)), and it doesn't have the major issues that the previous approach had, while it's still a major boon to console players.

I'll leave an approve and just respond to the comment here:

- Yes, hats getting stuck in the air in general is an issue already, but my major gripe with it here was that because *this* new hitbox is not parented to the player *model* (which ragdolls away), the hat-in-midair situation gets exacerbated. Parenting this one may require adding it to the hitboxes to disable in the ragdoll component. Fixing the whole hat issue is indeed way out of scope for this PR.
- I mentioned magnetism as a possible *alternative*, and I think we could test how this assist fares in demo sessions before implementing other approaches (this could also tell us how much of a balance difference this aim assist makes)
- I know conecasts aren't a thing here, which is why I (inadequately) added (approximated) in front to mean e.g. using differently-sized spherecasts :sweat_smile:
- Yeah, close range may be quite difficult as well, so let's just see how this distance-agnostic solution fares :rocket:

**Pull request merged**